polytracker/src/dfsan_rt/dfsan/polytracker_tracing.cpp

```diff
@@ -0,0 +1,86 @@
+#include <iostream>
+#include <mutex>
+#include <sstream>
+
+#include "polytracker/tracing.h"
+
+namespace polytracker {
+
+size_t numTraceEvents = 0;
+std::mutex traceEventLock;
+
+const std::list<dfsan_label> Trace::EMPTY_LIST = {};
+
+TraceEvent::TraceEvent() : previous(nullptr), next(nullptr) {
+  traceEventLock.lock();
+  eventIndex = numTraceEvents++;
+  traceEventLock.unlock();
+};
+
+std::string BasicBlockTrace::str() const {
+  std::stringstream s;
+  s << fname << " @ BB" << index.index() << " #" << entryCount;
+  return s.str();
+}
+
+bool FunctionCall::consumesBytes(const Trace& trace) const {
+  if (mConsumesBytes != CachedBool::UNKNOWN) {
+    return mConsumesBytes == CachedBool::TRUE;
+  }
+  for (TraceEvent* event = ret; event; event = event->previous) {
+    if (event->eventIndex <= this->eventIndex) {
```

Comment:
I think I mentioned this on this PR earlier or the other PR, but I'm not sure what the <= operation means for traces, or specifically events. 

I believe this function determines if a specific function call "consumes" bytes from the input stream by checking if they were the last to use it. 

Is this function also recursive? specifically line 51 bb->function->consumeBytes. 

Is the goal of this to see if a function consumes ANY bytes, or specific ones? I guess True and False are for any? Sorry for so many questions, trying to make sure I understand everything

Reply:
This function is effectively how I do the slicing, and determine what can be omitted from the output JSON. When we are recording `FunctionCall` and `FunctionReturn` trace events, we automatically assign member pointers to-and-from each other. `eventIndex` is a unique index for the event that just gets iterated on every event. So if `this->eventIndex == 9`, then `this` is the 10th event in the trace. The indexes are also unique across threads (we just increment the global counter inside of a mutex lock). Here is basically how this function works:

1. We want to determine whether the function executed in `this` event consumed any tainted bytes
2. Jump to the associated `FunctionReturn` event for this function call (which is `this->ret`)
3. Iteratively walk backwards through the trace. If we encounter a `BasicBlockEntry`, check if that BB processed tainted bytes. If so, we are done.
4. If we encounter another `FunctionReturn` event, then that is associated with a function call that was called within this function. So get the associated `FunctionCall` event for that return (`ret->call`) and recursively call `consumesBytes()` on that function call event. If it consumed bytes, we are done. If not, continue backward from there.
5. If we get all the way back to the initial `FunctionCall` without having found any tainted bytes, then we return `false`
6. If we somehow encounter an event with an index _before_ `this->eventIndex`, then that means we had some sort of instrumentation error that resulted in a `FunctionCall` not being properly associated with a `FunctionReturn`. (That's the purpose of the `eventIndex` comparisons)

We finally memoize the result so we don't do any extra work during the recursive calls.